### PR TITLE
[onert] Add shape inference for reshape with new_shape option

### PR DESCRIPTION
**Pull request by periannath**
- Shape inference of reshape uses new_shape option
  - If new shape is given by second input tensor, use that tensor as new shape
  - If above condition is not met, use new_shape option as new shape

ONE-DCO-1.0-Signed-off-by: JiHwan Yeo <jihwan.yeo@samsung.com>

---

For issue : #2393

### CODE REVIEW

**Comment by glistening** [comment]
@periannath Do you have test that you'd like to solve from #2393 and this PR?

**Comment by periannath** [comment]
@glistening I've created TFLite file from `res/TensorFlowLiteRecipes/Reshape_000/test.recipe`. It has one input tensor and get `new_shape` as option.

File: runtime/onert/core/src/util/shapeinf/Reshape.cc [hunk 1]
```diff
@@ -75,25 +75,41 @@ void StaticInferer::visit(const ir::operation::Reshape &op)
...
+    const auto shape_idx{op.getInputs().at(ir::operation::Reshape::Input::SHAPE)};
+    const auto &shape = _operands.at(shape_idx);
 
-  // Let's check the second input
-  const auto shape_idx{op.getInputs().at(ir::operation::Reshape::Input::SHAPE)};
-  const auto &shape = _operands.at(shape_idx);
-
-  ir::Shape new_shape;
-  if (shape.isConstant())
+    if (shape.isConstant())
```
Comment (ragmani) [inline runtime/onert/core/src/util/shapeinf/Reshape.cc:85]: Does `shape` operand have a higher priority than `op.param().new_shape`?

File: runtime/onert/core/src/util/shapeinf/Reshape.cc [hunk 2]
```diff
@@ -75,25 +75,41 @@ void StaticInferer::visit(const ir::operation::Reshape &op)
     return;
   }
 
-  if (op.getInputs().size() == 1)
+  // New shape is given by second input tensor
+  if (op.getInputs().size() == 2)
```
Comment (ragmani) [inline runtime/onert/core/src/util/shapeinf/Reshape.cc:79]: Does shape operand have a higher priority than op.param().new_shape?

Comment (glistening) [reply runtime/onert/core/src/util/shapeinf/Reshape.cc:79]: @ragmani Yes. According to tflite's `reshape` implementation, `input` has higher priority than `param` though it is different to https://www.tensorflow.org/lite/guide/ops_compatibility. As I understand, `tflite` seems to use `param` when the shape is determined before runtime, it uses `input` for dynamic shape. If `reshape` has `input`, `input` should be handled with high priority.

**Comment by glistening** [comment]
@periannath Do you have the test in your pc? If you're not going to run in CI for some reason, could you please attach the test in comment?

**Comment by periannath** [comment]
@glistening Here is the generated tflite file. [reshape_000.zip](https://github.com/Samsung/ONE/files/4816197/reshape_000.zip)

**Pull request merged**